service/javadsl/integration-client/src/main/java/com/lightbend/lagom/javadsl/client/integration/LagomClientFactory.java

```diff
@@ -167,10 +168,25 @@ private LagomClientFactory(EventLoopGroup eventLoop, WSClient wsClient, WebSocke
     public void close() {
         closeGracefully(wsClient::close);
         closeGracefully(webSocketClient::shutdown);
-        closeGracefully(actorSystem::terminate);
+        closeGracefully(this::coordinatedShutdown);
         closeGracefully(() -> eventLoop.shutdownGracefully(0, 10, TimeUnit.SECONDS));
     }
 
+    private void coordinatedShutdown() throws InterruptedException, ExecutionException, TimeoutException {
+        // CoordinatedShutdown may be invoked many times over the same actorSystem but
+        // only the first invocation runs the tasks (later invocations are noop).
+        Optional<String> runFromPhase =
+            OptionConverters.toJava(CoordinatedShutdownProvider.loadRunFromPhaseConfig(actorSystem));
+        CoordinatedShutdown cs = CoordinatedShutdown.get(actorSystem);
```

Comment:
How are instances of this class created? Shouldn't we inject CoordinatedShutdown just like we did for the actor system?

Reply:
Good call. This code could be split in two:

 * the creation of the `CoordinatedShutdown` instance (would have to be moved into `public static LagomClientFactory create`below)
 * the actual `run` of `cs`.